### DISCUSSION TITLE
How to determine that the input text occurred after clicking on the inline button in telegram bot?

### DISCUSSION THREAD

**Discussion by serjo96** [nksmnf/nestjs-telegraf#1083] [Q&A]
I have such a user case for a telegram bot:
1) press /menu, get inline menu
2) In the menu, click on the feedback button
3) enter the feedback in the text and inform that we have accepted the feedback
4) return the next menu message

I still don't get how to determine in text handler that moment with the message sent only after pressing the button, and not in the arbitrary text entered at any time.

**Comment by ramincsy** [comment]
Use per-user (or per-chat) state when handling the callback. Set a flag in the session from the inline button’s `callback_data`, then consume it in the next text update:

```ts
import { Action, Ctx, On, Update } from 'nestjs-telegraf';
import { Context } from 'telegraf';

type Session = { awaitingFeedback?: boolean };
type BotContext = Context & { session: Session };

@Update()
export class FeedbackUpdate {
  @Action('feedback')
  async startFeedback(@Ctx() ctx: BotContext) {
    await ctx.answerCbQuery();
    ctx.session.awaitingFeedback = true;
    await ctx.reply('Please send your feedback as a text message.');
  }

  @On('text')
  async receiveFeedback(@Ctx() ctx: BotContext) {
    if (!ctx.session.awaitingFeedback) return;

    ctx.session.awaitingFeedback = false; // consume the one expected reply
    const text = 'text' in ctx.message ? ctx.message.text : '';
    // save/process text here
    await ctx.reply(`Thanks for the feedback: ${text}`);
  }
}
```

The inline button should use `callback_data: 'feedback'`. Ensure a session middleware (for example `telegraf-session-local` or your chosen store) is installed before the handlers. The callback update and the later message are separate Telegram updates, so a flag/scene is what distinguishes the expected feedback from arbitrary text; a wizard scene is another good fit for a multi-step form.
